README.txt
PyTivo Video Manager - README.txt

see changelog.txt for a history of changes

what it is
==========

PyTivo video manager (vidmgr) is an HME application that allows you to request that videos in your 
library be "Pushed" to your tivo instead of the normal pull.  Since it's an HME app, this request
can be made from your easy chair with your tivo remote.  Pushing allows videos that are in a
compatible MP4 format to be transferred as is - saving time and space - instead of the transcoding
that a pull will always cause (although a push can also be requested for a video format that requires
transcoding).  vidmgr can also be used to delete videos from your library.

vidmgr was recently upgraded to support shares of type dvdvideo.  Deletion of these videos is not 
supported.

vidmgr is a TiVo HME application designed to operate under wmcbrine's hme for python
framework.  It is NOT a stand-alone application.  Please install the pyhme package (I have tested with version 0.19)
and make sure it is running before you install vidmgr.

Vidmgr is designed to operate in conjunction with the PyTivo transcoding server - If you are not 
using pytivo, then there is no point in installing vidmgr.  I have tested vidmgr with the mcbrine
fork of pytivo.  I am not familiar with the other forks to know whether or not it will run there.

====================================================================================================

installation/setup
==================

If you've gotten this far, you must want to install this.  It's very simple:

1. Go to the directory where you have pyhme installed.  There should be a lot of subdirectories here,
one per application.  Create a new directory here named "vidmgr"

2. Copy the contents of the vidmgr directory tree from this package into this new directory.  This should
include several .py files and a directory named skins.  Within the skins directory, there will be a
directory named orig and a directory named blue.  Each of these will have a bunch of .png files.   BTW,
metadata.py is an extract of the same named file that is part of pytivo.  I can't take the credit for this code.

3. Copy config.merge from this package into the pyhme main directory. 

4. Configure - you will need to merge the config.merge file that was delivered with this package with
your config.ini file that you are currently using.  There are four areas you need to pay attention to:  

	a) You may or may not have an "apps" line under the heading [hmeserver].  If you do not, then
the hmeserver will start all apps that it finds.  If you do, then it will only start the named apps.
So if you do have this line and you do want to run vidmgr, add the word "vidmgr" to this line - no quotes
or commas or other punctuation.

	b) you can specify various vidmgr options by putting entries in the [vidmgr] section of the config.ini
as follows:

[vidmgr]
exts=.mpg .mp4 .avi .wmv .m4v

   this names the file extensions you are interested in
      
descsize=20
   this gives the font point size that will be used for the description text.  20 is the default
   
skin=name
   this is the name of the directory under skins that contains all of the png files that are used
   to draw the screen.  default is "orig", although the package is also shipped with a "blue" skin
   
deleteallowed=true
   this determines whether or not deletion of videos is permitted.  Default is true, set to false
   if you do not want this capability
   
display=value
   determines what information is displayed about videos in the various lists.  allowable values are
   episodetitle  - displays only the episode title
   episodenumtitle - displays the episode number followed by the title
   file - simply displays the filename
   normal - (default) displays program title followed by episode title
   
sort=value
   determines how listings are sorted:
   episodenumber - sort based on episode number - note this is a string sort - not a numeric sort
   file - sort based on the filename
   normal - (default) sorts based on the program title and episode title

metafirst = title seriesTitle episodeTitle description
metaignore = isEpisode isEpisodic
   these two items determine which metadata is displayed first in the info screen and which is ignored. 
   Spelling and case are significant - the name must match exactly.  The default values are those
   values given above
metaspace = name name
metaspaceafter = name name
metaspacebefore = name name
   determines that there should be a blank line in the display before or after the indicated metadata items.  The
   default is an empty list so there will be no blank lines.  metaspace and metaspaceafter are synonyms
metamergefiles = False or True (default = True)
metamergelines = False or True (default = False)
   If there are multiple metafiles that correspond to a video file, these two options control how the data is to be merged.
   metamergefiles = False indicates that the files are not to be merged at all - only the more specific file is to be used.
   If metamergefiles if True, the default value, data from the less specific files is over-written/replaced with data from more
   specific files depending on the value for metamergelines.  If metamergelines is False (the default) then a repeating metadata
   key will REPLACE any previous value read.  If it is true, the new data will be concatenated to the old value separated
   by a space.  Note that metamergelines has NO effect on metadata items that start with a 'v' (vActor, etc).  These data
   items will continue to be processed as arrays.
   
   Metadata files are searched for/processed in the following order:
   	1) .meta/default.txt
   	2) default.txt
   	3) .meta/<title>.txt
   	4) <title>.txt
   Where <title> is the base name of the video file - or "folder" for directories.  DVDVideo shares have a few other
   quirks concerning metadata - see below.
   
infolabelpercent=30
   specifies the width, in percentage of the label field on the info screen.  Default is 30, but I have found that 
   20 works well for HD screens
   
inforightmargin=20
   specifies the width, in pixels of pad area on the right side of the info screen.  Default is 20. 0-100 allowed

thumbjustify=left
   specifies how thumbnail images should be justified.  default = left, can be center or right


	c) You need to tell vidmgr about your Tivos.  For each tivo, you need to specify the name and
the TSN.  The format for this is:
[tivos]
tivo1.name=Family Room
tivo1.tsn=TSN1
tivo2.name=Master Bedroom
tivo2.tsn=TSN2

You can have an arbitrary number of Tivos, but as soon as vidmgr detects a gap in the numbering
sequence it will stop parsing.  Make sure the TSN's are accurate as this is how pytivo knows which
tivo to push to.

	d) You need to tell vidmgr about your PyTivo instances.  There are 4 possible pytivo parameters:
		- config is mandatory and is the fully qualified name of the pytivo config file. 
vidmgr reads this file to determine the share names and locations.  
		- You may specify an ip address for the machine on which this instance of pytivo is running.
If you do not specify one, the local IP address is used.
		- If the pytivo config file names a port in the server section, then vidmgr will
use that port number.  Otherwise you need to specify the port number here.
		-Finally, if your hme server is running in a different host environment than this
instance of pytivo, then you need to specify the directory separator character for the pytivo environment.

format for specifying pytivo information:
[pytivos]
pytivo1.config=/usr/local/pytivo/pyTivo.conf
pytivo1.ip=192.168.1.201
pytivo1.port=9032
pytivo1.sep=/

You can have an arbitrary number of pytivos, but as soon as vidmgr detects a gap in the numbering
sequence it will stop parsing.

A note about the separator:  If you are running both vidmgr and pytivo on the same machine, then this
is not required.  However, if (as was happening while I was developing) you are running vidmgr
in a Windows environment (where the directory separator is unfortunately a backslash '\') and 
you are running pytivo in a linux environment (where the separator is a forward slash '/') then
you need to specify "pytivox.sep=/".  Otherwise, vidmgr will happily send its requests to 
pytivo using a backslash in the paths and this will cause pytivo to choke.

=======================================================================================================

Usage
=====

Vidmgr presents a directory tree to you.  You can step into and out of directories using the normal
tivo navigation keys.  The directory tree is rooted at the list of shares, unless there is only 1 share
in which case it is rooted at the topmost directory of that share.

while on a list of video files, you can also navigate using the number keys.  1 takes you 10% of the
way through the list, 5 = 50%, 7 = 70%, etc.  0 alternately takes you to the end of the list and then to the
beginning of the list

In HD mode, vidmgr will also show video artwork on the right hand side of the screen.  Vidmgr looks for
the following file:  <full-video-file-name-including-extension>.jpg or, if this doesn't exist, folder.jpg.
The view into which this graphic is placed is 620 pixels wide by 450 pixels high.  If your graphic exceeds
those dimensions it will be scaled while maintaining the aspect ratio.  Folder.jpg will also be the thumbnail
used for the enclosing folder or share.  Also, if there is a folder.txt file in a directory, or in the
subtending .meta directory, its contents - notably the description field will be shown on the display
above the thumbnail.

For dvdvideo shares, vidmgr is totally dependent on accurate metadata.  Metadata (and thumbnails) all belong
in the directory containing the VIDEO_TS directory or in a subtending .meta directory.  Metadata is processed
as follows.  default.txt contains the DVD metadata.  __Txx.mpg.txt contains the metadata for title xx.  The
title-specific metadata is overlaid on top of the DVD metadata according to the metamergefiles and metamergefiles
configuration parameters.  The thumbnail for the DVD is in a file named folder.jpg.  In addition, it is
possible to have a thumbnail for a specific title;  the name should be __Txx.mpg.jpg.  If there is no metadata
for a dvdvideo share, vidmgr assumes that there is only 1 title, and its title is the directory name.

Once you choose a video file, you will be shown some of the metadata associated with that file,
and then have two options - push or delete (delete can be disabled - see configuration above).  In
HD mode, this detail will be on the right side of the screen as you navigate through the directories - it
is not on a separate screen

If you choose delete, you will be asked to press thumbs-up to confirm and when you do, the file and
its associated metadata file will be deleted.  If you press ANYTHING other than thumbs-up, the
delete is cancelled.

If you choose push and you only have 1 tivo, then it will simply initiate a push to that tivo and
give you a confirmation message.  If you have multiple tivos, you need to choose the one you want to
push to from the provided list.  After choosing, vidmgr will initiate the push and give you the
confirmation message.  The confirmation message can be dismissed with ANY keypress

At any time on any list or on a details screen, you can press the info button to see a complete
list of the metadata.  You can control which metadata items appear at the front of this display
and which are ignored (see configuration above).  On the info display, you can press left, clear, or info
to return to the screen you came from.  If the information does not fit on one page, you will see
paging cues and you can use either up/down or channel up/down to traverse the pages.


changelog.txt
PyTivo Video Manager - change log
Version 0.7c - 8/2/2011
- improved logic when looking for dvd titles - it does not stop on title gaps anymore
- added options to control merging of metadata files: metamergefiles indicates if files should be merged, if files are 
  to be merged, metamergelines indicates if lines in one file should be merged with or replace lines from previous files.

Version 0.7b - 7/20/2011
- bug fix from yesterday

Version 0.7a - 7/19/2011
- allow pop-up information window to be displayed for folders if they have metadata.  ALso be more defensive if icons are missing

Version 0.7 - 7/11/2011
- added support for dvdvideo pytivo shares/plugin.  See readme file for details

Version 0.6b
- added image scaling back in - but only for images that are too big for the view

Version 0.6a 6/17/11
- put a try block around the stat operation in metadata.basic to catch broken links and other errors.  The action
is to simply put an error message into the metadata description field
- made use of the cleanup method as the correct place to save the cache
- added support for folder/share metadata.  If a folder or share has a file named folder.txt (or .meta/folder.txt)
then the contents will be read as metadata, and any description field that is found will be displayed on the screen
above the thumbnail
- added inforightmargin to configuration file - the number of pixels of right margin to leave on the info screen, default=20

Version 0.6
- added thumbjustify option to indicate if thumbnail images should be right or center justified.  default = left

Version 0.5f 6/15/11
- removed PIL code and image scaling - increased the size of the thumbnail view to 620x450.  Now submenu and
tivomenu buttons can potentially appear over the thumbnail

Version 0.5e 6/15/11
- added a try block around PIL code i case jpg file cannot be processed

Version 0.5d 6/15/11
- added logic to load thumbnails just in time rather that ahead of time.  This speeds up initial painting
of a page, but slows down navigation within the page.  Beacuse of this, 
- I added an image cache.  If you exit the app cleanly - by hitting left all the way out - the cache is saved to disk
for future runs.  The cache got too big to save to disk with every change, so this was a compromise.  The cache uses lru
logic to keep the last X used images in memory.   I had to choose some kind of limit here - I chose 100 for now - we'll
see how that works out.  Also, the cache exists at the module level, but yet is accessed, and possible changed, at the
HME thread level.  Because of this, I had to mutex exclusion around parts of the cache logic.
- added logic to scale the thumbnail to fit in the 320 pixel wide view frame.  Thanks to wmcbrine for the picture
sample hme app that told me how to do this.  Side effect of this - you need to have the python image lib (PIL) installed
- widened the right hand margin on the info page a bit

Version 0.5c 6/13/11

- added ability to jump a percentage of the way through long lisys by pressing digit keys - 5 = 50%, etc.
0 key behaves like advance - alternates between taking you to end and to beginning.
- expanded metaspace feature from 0.5b to have metaspacebefore and metaspaceafter.  If you add metadata items
to these lists, they will have a blank line before/after them in the info display.  The logic will NOT
display 2 blank lines in a row, nor will if display a blank on the first line of the display.


Version 0.5b 6/13/11

- extended the display of thumbnails for directories to include shares as well
- added metaspace to the config file to indicate if a blank line should be added to the display after
a particular item
- added infolabelpercent to configuration to specify the width of the label on the info screen as a percent

Version 0.5a 6/12/11

moved metaignore and metafirst into config file so you can specify which metadata you would like
to ignore and which you would like to see first

Added logic to show folder.jpg as the thumbnail for a directory if it exists in that directory or in the
subtending .meta directory

Version 0.5

-added the ability to specify deleteallowed=False in the config file to prevent file deletions
-formalized the skin concept.  Now skins are kept in separate directories under the skins directory, and
the directory you want to use for a skin is named in the config file e.g. skin=blue.  ALso, changed the
image loading logic such that if the HD version of an image does not exist, then it will use the SD version
of that file.  This solved the situation where there were duplicate files just to satisfy a naming convention.
- moved metadata.py into the vidmgr directory.  It didn't belong up in the pyhme directory and I never liked
having it up there.  I was too unfamiliar with python to know how to specify to look in the vidmgr subdirectory
to find the file (it turns out I didn't need to do anything - python found it!!)
- split README.txt into README.txt and changelog.txt.  changelog.txt will contain the change history, and README
will be maintained as the always current description of the application
- fixed the positioning bug in 0.4c where the "Push" button was over top of the bottom line of the description

Version 0.4c

Adjusted thumbnail position and calculated description height based in descsize so that only full lines appear

Version 0.4b

Increased list size on HD screen from 12 to 14

Version 0.4

Added an info page that pops up when you press the info button and displays all the metadata about the current video

Version 0.3c

Fixed bug when navigating OUT of an empty directory and bug where share list was not being updated when a video is deleted

Version 0.3b

Added 1) share and directory listings show the number of videos and directories contained within them, and 2) added a "Please Wait" icon
during length operations - currently this includes:  building a directory listing, pushing a video, and deleting a video.

Version 0.3a


vidmgr/thumbcache.py
'''
Created on Jun 14, 2011

@author: jbernard
'''
import cPickle as pickle
import os
import thread
import Image as img
from cStringIO import StringIO

CACHEFILE = 'thumbs.cache'

def resizePic(fn, width, height):
	try:
		pic = img.open(fn)
	except:
		return None
	
	pic.draft('RGB', (width, height))
	if pic.mode == 'P':
		pic = pic.convert()
	filew, fileh = pic.size

	if ((filew > width) or (fileh >height)):
		# determine horizontal and vertical ratios of pic to view
		ratiow = filew / float(width)
		ratioh = fileh / float(height)
		# choose the larger of the 2 ratios
		ratio = ratiow
		if ratioh > ratiow:
			ratio = ratioh
		# and scale accordingly
		nheight = int(fileh / ratio)
		nwidth = int(filew / ratio)
		pic = pic.resize((nwidth, nheight), img.ANTIALIAS)
		
	out = StringIO()
	pic.save(out, 'JPEG')
	encoded = out.getvalue()
	out.close()
		
	return encoded
	
class ThumbCache:
	def __init__(self, dir, size, vwidth, vheight):
		self.filename = os.path.join(dir, 'thumbs.cache')
		self.cacheChanged = False
		self.lrumap = []
		self.maxSize = size
		self.width = vwidth
		self.height = vheight
		self.mutex = thread.allocate_lock()
		
		# try to load the cache file if it exists
		self.cache = {}
		
		if (os.path.exists(self.filename)):
			print "Loading thumbnail cache"
			
			try:
				f = open(self.filename)
			except:
				print "Error opening thumbnail cache file"
			else:
				try:
					self.cache = pickle.load(f)
				except:
					print "Error loading thumbnail cache"
					self.cache = {}
					return
					
			for k in self.cache:
				self.appendMap(k)
			self.adjustMap()
			print str(len(self.cache)) + " thumbnails loaded from cache"
		else:
			print "Cache file does not exist - no thumbnails loaded"
	
	def sortMap(self):
		def cmplru (left, right):
			if (left[0] > right[0]): return 1
			if (left[0] < right[0]): return -1
			return 0
		
		self.lrumap = sorted(self.lrumap, cmp=cmplru)
	
	def adjustMap(self):
		while len(self.lrumap) > self.maxSize:
			(lru, key) = self.lrumap[0]
			del self.cache[key]
			del self.lrumap[0]
		i = 0
		for m in self.lrumap:
			m[0] = i
			i += 1
			
	def appendMap(self, key):		
		self.lrumap.append([self.maxSize, key])

	def addMap(self, key):
		self.appendMap(key)
		self.adjustMap()

	def updateMap(self, key):
		for m in self.lrumap:
			if m[1] == key:
				m[0] = self.maxSize # force to have the highest possible lru value
				break
		self.sortMap()
		self.adjustMap()

	def getImageData(self, filename):
		mt = 0
		try:
			mt = os.path.getmtime(filename)
		except os.error:
			# file does not exist
			return None
		
		self.mutex.acquire()

		adding = False
		if filename in self.cache:
			if self.cache[filename]['mtime'] >= mt:
				# don't load the file - just use the cache
				self.updateMap(filename)
				self.mutex.release()
				return self.cache[filename]['data']
			
			# file is in the cache but out of date
		else:
			# file is not in the cache - add it
			adding = True
			
		# either file is newer than cache
		# or file is NOT yet in the cache
		# load it in either case
		pdata = resizePic(filename, self.width, self.height)
		if pdata == None:
			self.mutex.release();
			return None

		if adding:
			self.cache[filename] = {}
		
		self.cache[filename]['data'] = pdata
		self.cache[filename]['mtime'] = mt

		self.cacheChanged = True
		if adding:
			self.addMap(filename)
		else:
			self.updateMap(filename)

		self.mutex.release()
		return self.cache[filename]['data']
	
	def hasChanged(self):
		return self.cacheChanged
	
	def size(self):
		return len(self.cache)
	
	def __str__(self):
		return str(self.lrumap)
	
	def saveCache(self):
		if self.hasChanged():
			self.mutex.acquire()
			try:
				f = open(self.filename, 'w')
			except:
				print "Error opening thumbnail cache file for write"
			else:
				try:
					pickle.dump(self.cache, f)
				except:
					print "Error saving thumbnail cache"
				else:
					f.close()
					self.cacheChanged = False
			self.mutex.release()
		

vidmgr/metadata.py
#!/usr/bin/env python

import os
from datetime import datetime

# Something to strip
TV_RATINGS = {'TV-Y7': 'x1', 'TV-Y': 'x2', 'TV-G': 'x3', 'TV-PG': 'x4', 
			  'TV-14': 'x5', 'TV-MA': 'x6', 'TV-NR': 'x7',
			  'TVY7': 'x1', 'TVY': 'x2', 'TVG': 'x3', 'TVPG': 'x4', 
			  'TV14': 'x5', 'TVMA': 'x6', 'TVNR': 'x7',
			  'Y7': 'x1', 'Y': 'x2', 'G': 'x3', 'PG': 'x4',
			  '14': 'x5', 'MA': 'x6', 'NR': 'x7', 'UNRATED': 'x7'}

MPAA_RATINGS = {'G': 'G1', 'PG': 'P2', 'PG-13': 'P3', 'PG13': 'P3',
				'R': 'R4', 'X': 'X5', 'NC-17': 'N6', 'NC17': 'N6',
				'NR': 'N8', 'UNRATED': 'N8'}

STAR_RATINGS = {'1': 'x1', '1.5': 'x2', '2': 'x3', '2.5': 'x4',
				'3': 'x5', '3.5': 'x6', '4': 'x7',
				'*': 'x1', '**': 'x3', '***': 'x5', '****': 'x7'}

HUMAN = {'mpaaRating': {'G1': 'G', 'P2': 'PG', 'P3': 'PG-13', 'R4': 'R',
						'X5': 'X', 'N6': 'NC-17', 'N8': 'Unrated'},
		 'tvRating': {'x1': 'TV-Y7', 'x2': 'TV-Y', 'x3': 'TV-G',
					  'x4': 'TV-PG', 'x5': 'TV-14', 'x6': 'TV-MA',
					  'x7': 'Unrated'},
		 'starRating': {'x1': '1', 'x2': '1.5', 'x3': '2', 'x4': '2.5',
						'x5': '3', 'x6': '3.5', 'x7': '4'}}

BOM = '\xef\xbb\xbf'

def tag_data(element, tag):
	for name in tag.split('/'):
		new_element = element.getElementsByTagName(name)
		if not new_element:
			return ''
		element = new_element[0]
	if not element.firstChild:
		return ''
	return element.firstChild.data

def _vtag_data(element, tag):
	for name in tag.split('/'):
		new_element = element.getElementsByTagName(name)
		if not new_element:
			return []
		element = new_element[0]
	elements = element.getElementsByTagName('element')
	return [x.firstChild.data for x in elements if x.firstChild]

def _tag_value(element, tag):
	item = element.getElementsByTagName(tag)
	if item:
		value = item[0].attributes['value'].value
		name = item[0].firstChild.data
		return name[0] + value[0]

def from_text(full_path, mergefiles=True, mergelines=False):
	metadata = {}
	path, name = os.path.split(full_path)
	title, ext = os.path.splitext(name)

	for metafile in [os.path.join(path, title) + '.properties',
					 os.path.join(path, '.meta', 'default.txt'),
					 os.path.join(path, 'default.txt'),
					 os.path.join(path, '.meta', name) + '.txt',
					 full_path + '.txt']:
		if os.path.exists(metafile):
			if not mergefiles:
				metadata = {}
			prevkeys = metadata.keys()
			sep = ':='[metafile.endswith('.properties')]
			for line in file(metafile, 'U'):
				if line.startswith(BOM):
					line = line[3:]
				if line.strip().startswith('#') or not sep in line:
					continue
				key, value = [x.strip() for x in line.split(sep, 1)]
				if not key or not value:
					continue
				value = ''.join([x for x in value if ord(x) < 128])
				if key.startswith('v'):
					if key in metadata:
						metadata[key].append(value)
					else:
						metadata[key] = [value]
				else:
					if key in metadata:
						if key in prevkeys and mergelines:
							metadata[key] += ' ' + value
							prevkeys.remove(key)
						elif key in prevkeys:
							metadata[key] = value
							prevkeys.remove(key)
						elif mergelines:
							metadata[key] += ' ' + value
						else:
							metadata[key] = value
					else:
						metadata[key] = value

	for rating, ratings in [('tvRating', TV_RATINGS),
							('mpaaRating', MPAA_RATINGS),
							('starRating', STAR_RATINGS)]:
		x = metadata.get(rating, '').upper()
		if x in ratings:
			metadata[rating] = ratings[x]

	return metadata

def basic(full_path):
	base_path, name = os.path.split(full_path)
	title, ext = os.path.splitext(name)
	try:
		mtime = os.stat(full_path).st_mtime
	except:
		desc = "Error trying to stat " + full_path
		mtime = 0
	else:
		if (mtime < 0):
			mtime = 0
		desc = ""
		
	originalAirDate = datetime.fromtimestamp(mtime)

	metadata = {'title': title,
				'description': desc,
				'originalAirDate': originalAirDate.isoformat()}

	return metadata
